### PR TITLE
#23175 adding secrets as velocity

### LINKED ISSUE #23175: Add a "Velocity Secrets" app and viewtool
When coding integrations in Velocity, you sometimes need to securely store Access Tokens, keys, urls, passwords etc for use in integrations\.  While we do not want to expose the whole "App" infrastructure to Velocity as that would be a security risk,  it would be nice to have a specific App \- called "VelocitySecrets" or something, that would be used to store and manage these "secrets"\.  
  
The  "VelocitySecrets" app would just have a key\/value field to which an administrator could add their secret tokens, ids and keys\.  Scripters and front end devs could then access these variables in a velocity script by calling `$dotsecrets.keyName`  

We can use the same permission pattern that we do for scripting privileges \- meaning we would check the last editor of the running velocity script calling the `$dotsecrets` to make sure they were a scripting user before outputting the secret\.

### PR DESCRIPTION
**Pull request by jdotcms**
The idea is to have
1) an app to write secrets for velocity
2) a view tool to access secrets from velocity

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/secrets/DotVelocitySecretAppConfig.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+        final DotVelocitySecretAppConfig config = DotVelocitySecretAppConfig.builder()
+                .withTitle(title)
+                .withExtraParameters(extraParameters).build();
+
+        DotVelocitySecretAppConfigThreadLocal.INSTANCE.setConfig(host.getIdentifier(), Optional.ofNullable(config));
+
+        return DotVelocitySecretAppConfigThreadLocal.INSTANCE.getConfig(host.getIdentifier());
+    }
+
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/secrets/DotVelocitySecretAppConfig.java:111]: Shouldn't we always use `AppSecrets` as the carrier of secrets and the call destroy once we're done? Here you're copying the secrets for convenience in strings and as they are immutable the reference sticks around in the VM.
I would try to stick to the  AppSecrets and always call destroy once it has been used.

**Review by nollymar — CHANGES_REQUESTED** [review]
Please update the branch. It seems to be outdated and the tests aren't running

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/secrets/DotVelocitySecretAppConfig.java (hunk 1 above)
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/secrets/DotVelocitySecretAppConfig.java:111]: Shouldn't we always use AppSecrets as the carrier of secrets and the call destroy once we're done? Here you're copying the secrets for convenience in strings and as they are immutable the reference sticks around in the VM.
I would try to stick to the AppSecrets and always call destroy once it has been used.

**Review by fabrizzio-dotCMS — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/secrets/DotVelocitySecretAppConfig.java (hunk 1 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/secrets/DotVelocitySecretAppConfig.java:111]: lemme think we I can do
May be I can destroy when the request is over

**Review by fabrizzio-dotCMS — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

**Pull request merged**